### PR TITLE
Avoid rewrapping modules with DDP/FSDP if already wrapped

### PR DESCRIPTION
**Pull request by ananthsub**
# What does this PR do ?
In case of multiple `trainer.fit` calls for the same LightningModule, it's possible the underlying module was wrapped with DDP/FSDP multiple times. This adds a check to ensure that for such cases, the module is wrapped only once.

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by farhadrgh** [comment]
@ananthsub The changes in [32cb34a](https://github.com/NVIDIA/NeMo/pull/12096/commits/32cb34a4733b2c8481979aecc44bef224ce7cbcf) plus an else statement to avoid unbound issue for `dist_module` worked for us and did not create `.module` prefixes every-time we call `trainer.fit()` in a loop.

```python
if HAVE_CUSTOM_FSDP and self.ddp_config.use_custom_fsdp and not isinstance(unwrapped_module, FullyShardedDataParallel):
    ....
elif not isinstance(unwrapped_module, DDP):
    ....
else:
    dist_module = unwrapped_module
```

File: nemo/lightning/fabric/plugins.py [hunk 1]
```diff
@@ -146,9 +146,10 @@ def convert_module(self, module: nn.Module) -> nn.Module:
             config = get_model_config(module.module)
             config.fp16 = self.dtype_config.fp16
             config.bf16 = self.dtype_config.bf16
-            if hasattr(module, 'module'):
+            # Avoid rewrapping the module if it's already of type Float16Module
+            if hasattr(module, 'module') and not isinstance(module.module, Float16Module):
                 module.module = Float16Module(config, module.module)
-            else:
+            elif not isinstance(module, Float16Module):
                 module = Float16Module(config, module)
```
Comment (farhadrgh) [inline nemo/lightning/fabric/plugins.py:153]: We need these changes to be applied to `MegatronMixedPrecision` as well:
https://github.com/NVIDIA/NeMo/blob/d23e1d614d1fc6f80368482029eec69f5961da37/nemo/lightning/pytorch/plugins/mixed_precision.py#L154-L157

Comment (farhadrgh) [reply nemo/lightning/fabric/plugins.py:153]: however, even with this change we are wrapping with `Float16Module` again. The hierarchy of modules printed just before the `convert_module()` method in `MegatronMixedPrecision `:

1st run
```
BionemoLightningModule(
   (module): ESM2FineTuneSeqModel(
```

2nd run
```
BionemoLightningModule(
  (module): DDP(
    (module): Float16Module(
       (module): ESM2FineTuneSeqModel(
```
